sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,315 @@
+
+    def _update_medoid_ics_in_place(self, D, cluster_ics, medoid_ics):
+        """In-place update of the medoid indices"""
+
+        # Update the medoids for each cluster
+        for cluster_idx in range(self.n_clusters):
+
+            if sum(cluster_ics == cluster_idx) == 0:
+                warnings.warn("Cluster %d is empty!" % cluster_idx)
+                continue
+
+            # Find current cost that is associated with cluster_idx.
+            # Cost is the sum of the distance from the cluster
+            # members to the medoid.
+            curr_cost = np.sum(D[medoid_ics[cluster_idx],
+                                 cluster_ics == cluster_idx])
+
+            # Extract the distance matrix between the data points
+            # inside the cluster_idx
+            D_in = D[cluster_ics == cluster_idx, :]
+            D_in = D_in[:, cluster_ics == cluster_idx]
+
+            # Calculate all costs there exists between all
+            # the data points in the cluster_idx
+            all_costs = np.sum(D_in, axis=1)
+
+            # Find the index for the smallest cost in cluster_idx
+            min_cost_idx = np.argmin(all_costs)
+
+            # find the value of the minimum cost in cluster_idx
+            min_cost = all_costs[min_cost_idx]
+
+            # If the minimum cost is smaller than that
+            # exhibited by the currently used medoid,
+            # we switch to using the new medoid in cluster_idx
+            if min_cost < curr_cost:
+                # Find data points that belong to cluster_idx,
+                # and assign the newly found medoid as the medoid
+                # for cluster c
+                medoid_ics[cluster_idx] = \
+                    np.where(cluster_ics == cluster_idx)[0][min_cost_idx]
+
+    def transform(self, X):
+        """Transforms X to cluster-distance space.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape=(n_isamples, n_features)
+            Data to transform.
+
+        Returns
+        -------
+        X_new : {array-like, sparse matrix}, shape=(n_samples, n_clusters)
+            X transformed in the new space.
+        """
+
+        check_is_fitted(self, "cluster_centers_")
+
+        # Apply distance metric wrt. cluster centers (medoids),
+        # and return these distances
+        return self.distance_func(X, Y=self.cluster_centers_)
+
+    def predict(self, X):
+        """Predict the closest cluster each sample in X belongs to.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            New data to predict.
+
+        Returns
+        -------
+        labels : array, shape [n_samples,]
+            Index of the cluster each sample belongs to.
+        """
+
+        check_is_fitted(self, "cluster_centers_")
+
+        # Check that the array is good and attempt to convert it to
+        # Numpy array if possible
+        X = check_array(X)
```

Comment:
Sparse matrices should also be allowed here.
Please add `accept_sparse=['csr', 'csc']`